### PR TITLE
clarify how write_bytes can lead to UB due to invalid values

### LINKED ISSUE #330: When using `std::ptr::copy` and similar methods, does the memory need to be valid for T?
Plenty of methods in std::ptr take a generic T and use `*mut/const T`, but does this mean that the following code is UB?

```rust
let src: u8 = 42;
let mut dest: u8 = 10;
std::ptr::copy(&src as *const u8 as *const bool, &mut dest as *mut u8 as *mut bool);
```

What about copying a `u32` as a `(u16, u8)`, or a `(u16, u8, !)`, that doesn't uninit any padding, right?

Same question applies to `copy_nonoverlapping`, `swap`(`_nonoverlapping`), `write_bytes`. I don't see any other methods that this would apply to.

What about copying a pointer with an integer type, that preserves provenance?

---

I'd be inclined to say "no, it's not UB, the generic parameter is only used for its size and alignment, no other property of the type matters", but if that's the case, we should make that clear. It definitely seems to be *implied*.

### PR DESCRIPTION
**Pull request by RalfJung**
Fixes https://github.com/rust-lang/unsafe-code-guidelines/issues/330

Cc @5225225

### CODE REVIEW

**Comment by 5225225** [comment]
I'd include a code sample as well, personally.

```rust
let mut value: u8 = 0;
let ptr: *mut bool = &mut value as *mut u8 as *mut bool;
ptr.write_bytes(42, 1);
```

Something like that. Not sure.

And/or

> Additionally, note that changing `*dst` in this way can lead to undefined behavior later if the written bytes are not a valid representation of some `T`, but the act of writing the bytes is not undefined behavior.

That might be too redundant, but it is more explicit.

**Comment by eggyal** [comment]
To be clear, is it the subsequent read (of an invalid `T`) that's UB; or is writing the invalid `T` immediate UB even if not subsequently read?

**Comment by 5225225** [comment]
`write_bytes` doesn't write a T.

It could instead be written as

```rust
fn write_bytes(ptr: *mut u8, val: u8, count: usize, size: usize, align: usize) {
...
}
```

As in, the only thing it cares about is the size and align of T. So you could define write bytes as `write_bytes::<MaybeUninit<T>>` and nothing would change.

What the comment was warning about is just because write_bytes doesn't write a T, if you're writing to a `*mut T`, that strongly implies you will use the value *as* a `T` in the future, and that can be UB if it's invalid.

**Comment by RalfJung** [comment]
> I'd include a code sample as well, personally.

That's fair, I was just lazy...

> write_bytes doesn't write a T.

I agree with that comment but I have no idea if you are suggesting any docs change here.

**Comment by 5225225** [comment]
> I agree with that comment but I have no idea if you are suggesting any docs change here.

Was replying to a deleted comment, I'll hide it.

**Comment by RalfJung** [comment]
Ah... yeah @eggyal please don't delete comments, it's just confusing.

**Comment by eggyal** [comment]
Apologies.  I posted my comment before I saw https://github.com/rust-lang/rust/pull/99084#issuecomment-1179549566, which had already answered it; and then deleted within seconds not realising a reply was already being written... hadn't intended to cause all this confusion!

**Comment by saethlin** [comment]
Does this also close https://github.com/rust-lang/rust/issues/84184 ? Or will that issue just need an update?

**Comment by RalfJung** [comment]
I think that issue is basically https://github.com/rust-lang/unsafe-code-guidelines/issues/84, so I don't think it is entirely resolved by this PR.

This PR carefully uses a local variable of type `u8` and a pointer of type `bool` to avoid https://github.com/rust-lang/unsafe-code-guidelines/issues/84. Not that I expect anyone to actually deduce this from the example, but I wanted to avoid making a statement on https://github.com/rust-lang/unsafe-code-guidelines/issues/84 here. Or maybe that's just not possible? I don't think we are ready to decide about https://github.com/rust-lang/unsafe-code-guidelines/issues/84 though.

**Comment by saethlin** [comment]
You commented before that:

> I think it should be valid, but I don't think we have an RFC or a lang-team decision saying that it is valid. So we can't really recommend using it in the docs currently.

So I don't understand how if the example that already exists is on shaky ground, we can justify adding a second one. I can never tell if the docs are normative or not, but if they are we've already got the cart miles ahead of the horse here. What is the point of an RFC or a lang-team decision if this pattern is already enshrined as valid because it is in the docs?

**Comment by RalfJung** [comment]
The docs I am adding here do *not* use the potentially bad pattern.

**Comment by saethlin** [comment]
As far as I can tell the question that's up in the air in https://github.com/rust-lang/unsafe-code-guidelines/issues/84 is: Even if you don't "use" it, is setting the value to an invalid bit pattern in itself UB? If the docs are normative, this has already been decided and the answer is no.

Also, I'm not sure what it means to "use" a value. The reference defines "producing" a value but that's not the term you're using here and also that page in the reference has a big red warning that is is not exhaustive. Does taking a reference to a value use the value? It seems like the answer to that is already "no"... because the docs do it. Even though there has been a lot of discussion about this recently.

I understand what you intend for this example in the docs. My concern is that these docs already suggest that a whole bunch which seems to me is not settled has actually been decided, and this PR is adding to that situation.

**Comment by RalfJung** [comment]
>  If the docs are normative, this has already been decided and the answer is no.

Which part of the docs are you referring to?

All I want to do here is answer https://github.com/rust-lang/unsafe-code-guidelines/issues/330#issuecomment-1178725979 without making any other commitments. You seem to be saying that is not possible? I don't understand why, since https://github.com/rust-lang/unsafe-code-guidelines/issues/330#issuecomment-1178725979 is independent of https://github.com/rust-lang/unsafe-code-guidelines/issues/84.

File: library/core/src/intrinsics.rs [hunk 1]
```diff
@@ -2550,14 +2550,23 @@ pub const unsafe fn copy<T>(src: *const T, dst: *mut T, count: usize) {
...
-/// size_of::<T>()` bytes to the given region of memory results in a valid
-/// value of `T`. Using a region of memory typed as a `T` that contains an
-/// invalid value of `T` is undefined behavior.
-///
 /// Note that even if the effectively copied size (`count * size_of::<T>()`) is
 /// `0`, the pointer must be non-null and properly aligned.
 ///
+/// Additionally, note that changing `*dst` in this way can easily lead to undefined behavior (UB)
+/// later if the written bytes are not a valid representation of some `T`. For instance, the
+/// follwing is an **incorrect** use of this function:
```
Comment (saethlin) [inline library/core/src/intrinsics.rs:2558]:
```suggestion
/// following is an **incorrect** use of this function:
```

**Comment by saethlin** [comment]
Okay, phew. I think I've poorly communicated my concern here and I'm also re-familiarizing myself with the context. Sorry about that.

> Which part of the existing docs are you referring to?

https://github.com/rust-lang/rust/blob/268be96d6db196d79a1b2b8299a0485496e5a7ab/library/core/src/intrinsics.rs#L2580-L2608

> And which part of my PR makes any commitments regarding https://github.com/rust-lang/unsafe-code-guidelines/issues/84?

I do not think this PR makes any additional commitments. Sorry if I suggested that. I think we are already more committed by the example I linked above.

I'm concerned about the net effect of adding more documentation and more examples to `write_bytes`. I think that a user could be rightfully confused because there are two examples which seem to be demonstrating the same thing. I don't want readers to be guessing that there is some subtle distinction between these examples that has led to having two examples instead of just one. The things that stand out to _me_ as different is `Box` vs `bool` and one creates a reference to an invalid value where the other doesn't. I'm not sure we want a reader of these docs to be guessing or combing through UCG issues to figure out what's different.

If I had my way, we'd delete the existing example and just have the one you are trying to add with this PR.

**Comment by RalfJung** [comment]
> > Which part of the existing docs are you referring to?

Ah, I see.
No reference actually gets invalidated in the time where the `Box` is bad, so this is somewhere on the edge of UCG#84. I would also be fine with removing that example :shrug: 

I think it was actually meant to demonstrate the part of the docs this is editing, and that the UB is indeed delayed, not at the time of the write.

**Comment by RalfJung** [comment]
We haven't heard from @m-ou-se , so maybe let's re-roll the reviewer dice.

r? rust-lang/T-libs-api

**Comment by RalfJung** [comment]
Oh, did the syntax change?
@rustbot r? rust/lib-api

**Comment by RalfJung** [comment]
Yet another attempt at guessing the right syntax
r? libs

**Comment by RalfJung** [comment]
Filing an issue elsewhere to request review for a rustc PR? I am confused.^^

This changes the doc comment of a stdlib function. It doesn't change the syntax or types of any public API, but it changes / clarifies its contract. I'll leave it to you to decide what kind of review that needs.

**Comment by thomcc** [comment]
If it changes the contract you should make an [ACP](https://std-dev-guide.rust-lang.org/feature-lifecycle/api-change-proposals.html) about it.

If you want, you can wait until I review and I can tell you if I think this is needed, though.

**Comment by RalfJung** [comment]
Oh. If that much process is needed for this kind of PR nowadays I will probably have to reduce the amount of those PRs I write. I often notice small-ish issues in our unsafe API description, and like to write quick clarifications. But I don't have the time for a lot of extra paperwork here -- sorry.

**Comment by thomcc** [comment]
> Also that says ACP is for changes to unstable APIs. This is a change to the documentation of a stable API (like most of my stdlib documentation PRs).

Right, I guess the assumption is that we don't make changes to stable APIs.

---

Either way, this doesn't actually seem to change anything. This is a clarification of an existing API, and this was already UB in the same way, no?

**Comment by RalfJung** [comment]
We change the documentation of unsafe stable APIs all the time. It's very hard to get the wording just right, so this is a continuous process.

> Either way, this doesn't actually seem to change anything. This is a clarification of an existing API, and this was already UB in the same way, no?

Whether or not this changes anything depends on how you interpret the documentation. ;) My interpretation is that this just says, in a clearer way, what we already meant before.

**Comment by thomcc** [comment]
> Whether or not this changes anything depends on how you interpret the documentation. ;) My interpretation is that this just says, in a clearer way, what we already meant before.

I agree. The other option would be T-libs-API FCP, but I don't think this is needed (if someone disagrees they can r- or back it out) -- It's not *identical* to what was promised before, but I think the previous one *meant* to say something closer to this. I don't think that needs signoff, at least not in this case.

@bors r+ rollup

**Pull request merged**